core/src/main/scala/org/apache/spark/internal/Logging.scala

```diff
@@ -229,3 +237,33 @@ private[spark] object Logging {
     "org.slf4j.impl.Log4jLoggerFactory".equals(binderClass)
   }
 }
+
+private class SparkShellLoggingFilter() extends Filter {
+
+  /**
+   * If sparkShellThresholdLevel is not defined, this filter is a no-op.
+   * If log level of event is lower than thresholdLevel, then the decision is made based on
+   * whether the log came from root or some custom configuration
+   * @param loggingEvent
+   * @return decision for accept/deny log event
+   */
+  def decide(loggingEvent: LoggingEvent): Int = {
+    val thresholdLevel = Logging.sparkShellThresholdLevel
+    if (thresholdLevel == null) {
+      return Filter.NEUTRAL
+    }
+    val rootLevel = LogManager.getRootLogger().getLevel()
+    if (loggingEvent.getLevel().isGreaterOrEqual(thresholdLevel) ||
+      !loggingEvent.getLevel().eq(rootLevel)) {
```

Comment:
Doesn't just `!loggingEvent.getLevel().eq(rootLevel)` cover all cases?

If the log message level is different than the root level, then the only way it can happen is if the logger was customized.

If it's equal, then you need to check whether the logger's level was customized, do decide whether to filter based on the shell threshold.

(BTW the method comment doesn't fully explain the logic here.)

Reply:
So, the `loggingEvent.getLevel().isGreaterOrEqual(thresholdLevel)` is just an optimization. If the log message level is same as root level but it is greater than threshold level, then we still allow the event to pass through, irrespective of whether it was customized or not.

Reply:
That's my question. Doesn't `!loggingEvent.getLevel().eq(rootLevel)` already cover that?

The other check is only an optimization if `thresholdLevel` is the same as `rootLevel`, and in that case a better optimization would be to not set `thresholdLevel` (so the null check triggers instead).

Using the good old truth table:

```
Level       Thresh      Root      Action
INFO       WARN       INFO     Check for override
WARN     WARN       INFO     Allow
DEBUG   WARN       INFO     Allow
```

So the resulting action doesn't seem dependent on the threshold level at all, assuming it's always different than the root logger's level.

Reply:
Got it. I just checked and `thresholdLevel` is only set when it is different from `rootLevel`. So, this check is not needed and I will remove it.